pkg/asset/installconfig/installconfig.go

```diff
@@ -144,7 +144,7 @@ func (a *InstallConfig) finish(filename string) error {
 		a.AWS = aws.NewMetadata(a.Config.Platform.AWS.Region, a.Config.Platform.AWS.Subnets, a.Config.AWS.ServiceEndpoints)
 	}
 	if a.Config.Azure != nil {
-		a.Azure = icazure.NewMetadata(a.Config.Azure.CloudName)
+		a.Azure = icazure.NewMetadata(a.Config.Azure.CloudName, nil)
```

Comment:
Can you give some more details about how ARO will call `NewMetadata`? Will ARO replace the metadata set here? Or does ARO not load the install config asset?

Reply:
That is helpful. Thanks you.
Can you help me track down which calls in the ARO eventually lead to calls to `installConfig.Azure.Session`?

Reply:
I think I found it. It is through code such as https://github.com/Azure/ARO-RP/blob/58fd2f4b79b6d0a71fd2575db0cf77e18b722328/pkg/cluster/deploystorage.go#L179-L188. Right?

Reply:
Can we drill down into why you need the credentials in the installconfig metadata?
Are you using the metadata as a store for the credentials? Or are you using it to just so that the installer can have the credentials when generating assets?
Do you need the credentials marshaled? Or will the credentials only be used in the initial resolve of the graph and not later after reloading the graph?

Reply:
@staebler

> I think I found it. It is through code such as https://github.com/Azure/ARO-RP/blob/58fd2f4b79b6d0a71fd2575db0cf77e18b722328/pkg/cluster/deploystorage.go#L179-L188. Right?

You are right.

> Are you using the metadata as a store for the credentials? Or are you using it to just so that the installer can have the credentials when generating assets?

We only need creds so that the installer can generate assets.

> Do you need the credentials marshaled? Or will the credentials only be used in the initial resolve of the graph and not later after reloading the graph?

I believe we don't need them marshaled. I tried to avoid marshaling of the credentials in the earlier version of this PR ([here](https://github.com/openshift/installer/commit/52d40af64bc6acf13531de178cca33e40010f033#diff-da950a4ddabbede621d9d3b1058bb34f8931c89179306ee88a0e4d76a4cf0b13R48-R50) is one of the older commits). The earlier version works for ARO, but not for the installer because the installer reloads the graph and panics after unmarshalling because there will be no `credGetter` [here](https://github.com/openshift/installer/commit/52d40af64bc6acf13531de178cca33e40010f033#diff-6b7f83e796e074b2f55bd6a9b666d65d140fe7477bd98b78bcc51210980213ddR47).

In [our fork of the installer](https://github.com/openshift/installer/commit/afbdc9c2bec45e0b268e94ee31623712d93205eb#diff-6b7f83e796e074b2f55bd6a9b666d65d140fe7477bd98b78bcc51210980213ddR16) you can see that we we pass creds into the Metadata and do not have them marshalled as the field is not exported. It works for ARO, but in our fork backward compatibility with upstream installer is not the top priority :) I tried to rework this bit in a way that doesn't hurt the installer and doens't require gating code with an ARO-specific build tag.

I feel like you are asking these questions because you have an alternative aproach in mind. Please let me know if that's the caseю

Reply:
I do not have an alternative approach in mind, but I am trying to understand the problem more clearly so that I can consider whether there may be alternative approaches that are less likely to be accidentally removed in the future.